ocs_ci/ocs/external_ceph.py

```diff
@@ -0,0 +1,924 @@
+
+import paramiko
+from paramiko.ssh_exception import SSHException
+
+from ocs_ci.ocs.exceptions import CommandFailed
+
+logger = logging.getLogger(__name__)
+
+
+class Ceph(object):
+
+    def __init__(self, name='ceph', node_list=None):
+        """
+        Ceph cluster representation. Contains list of cluster nodes.
+        Args:
+            name (str): cluster name
+            node_list (ceph.utils.CephNode): CephNode list
+        """
+        self.name = name
+        self.node_list = node_list
+
+    def __eq__(self, ceph_cluster):
+        if hasattr(ceph_cluster, 'node_list'):
+            if all(atomic_node in ceph_cluster for atomic_node in self.node_list):
+                return True
+            else:
+                return False
+        else:
+            return False
+
+    def __ne__(self, ceph_cluster):
+        return not self.__eq__(ceph_cluster)
+
+    def __len__(self):
+        return len(self.node_list)
+
+    def __getitem__(self, key):
+        return self.node_list[key]
+
+    def __setitem__(self, key, value):
+        self.node_list[key] = value
+
+    def __delitem__(self, key):
+        del self.node_list[key]
+
+    def __iter__(self):
+        return iter(self.node_list)
+
+    def get_nodes(self, role=None, ignore=None):
+        """
+        Get node(s) by role. Return all nodes if role is not defined
+
+        Args:
+            role (str, RolesContainer): node's role. Takes precedence over ignore
+            ignore (str, RolesContainer): node's role to ignore from the list
+
+        Returns:
+            list: nodes
+
+        """
+        if role:
+            return [node for node in self.node_list if node.role == role]
+        elif ignore:
+            return [node for node in self.node_list if node.role != ignore]
+        else:
+            return list(self.node_list)
+
+    def get_ceph_objects(self, role=None):
+        """
+        Get Ceph Object by role. Returns all objects if role is not defined. Ceph object can be Ceph demon, client,
+        installer or generic entity. Pool role is never assigned to Ceph object and means that node has no Ceph objects
+
+        Args:
+            role (str): Ceph object's role as str
+
+        Returns:
+            list: ceph objects
+
+        """
+        node_list = self.get_nodes(role)
+        ceph_object_list = []
+        for node in node_list:
+            ceph_object_list.extend(node.get_ceph_objects(role))
+        return ceph_object_list
+
+    def get_ceph_object(self, role, order_id=0):
+        """
+        Returns single ceph object. If order id is provided returns that occurrence from results list, otherwise returns
+        first occurrence
+
+        Args:
+            role(str): Ceph object's role
+            order_id(int): order number of the ceph object
+
+        Returns:
+            CephObject: ceph object
+
+        """
+        try:
+            return self.get_ceph_objects(role)[order_id]
+        except IndexError:
```

Comment:
should we maybe print the exception in case `None` is returned?

Reply:
done